### PR TITLE
[Sign in] Load Terms of Service for remote data store

### LINKED ISSUE #746: [Terms of Service] Show terms of service after signin
When the user signin show the Terms of Service. Terms of Service would be fetched from the db itself user needs to accept that to use the application.

### PR DESCRIPTION
**Pull request by HemanParbhakar**
This branch is a part of Terms & Service Features that would be shown to user after SignIn. In this Branch the terms are getting fetched from the firestore db itself.

Fixes #746 

@atalekar PTAL?

### CODE REVIEW

**Comment by gino-m** [comment]
Hi @HemanParbhakar, please resolve conflicts, update branch, and fix broken unit tests. Thanks again!

File: gnd/src/main/java/com/google/android/gnd/MainViewModel.java [hunk 1]
```diff
@@ -151,7 +173,16 @@ void onSignedOut() {
 
   private void onSignedIn() {
     hideProgressDialog();
-    navigator.navigate(HomeScreenFragmentDirections.showHomeScreen());
+    if (termsState.getValue() == LoadState.NOT_FOUND) {
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/MainViewModel.java:176]: There's a race condition here: if the terms aren't loaded before the sign in completes `termsState.getValue()` will return null, and the user will be directed to the home screen. I think the right thing to do here would be to join into the stream that triggers onSignInState so that it doesn't call the observer until the terms are loaded.

File: gnd/src/main/java/com/google/android/gnd/model/Terms.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.android.gnd.model;
+
+import com.google.auto.value.AutoValue;
+
+/** Represents a single application user. */
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/model/Terms.java:21]: Description to be updated.

File: gnd/src/main/java/com/google/android/gnd/model/Terms.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+@AutoValue
+public abstract class Terms {
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/model/Terms.java:23]: `TermsOfService` might be more self-describing.

File: gnd/src/main/java/com/google/android/gnd/persistence/local/LocalValueStore.java [hunk 4]
```diff
@@ -72,4 +73,19 @@ public void saveMapType(int type) {
   public int getSavedMapType(int defaultType) {
     return preferences.getInt(MAP_TYPE, defaultType);
   }
+
+  /**
+   * This method will return whether the user have accepted the terms or not.
+   * @return true or false depending upon the user.
+   */
+  public Boolean areTermsAccepted() {
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/persistence/local/LocalValueStore.java:81]: Boolean -> boolean

File: gnd/src/main/java/com/google/android/gnd/persistence/remote/RemoteDataStore.java [hunk 5]
```diff
@@ -46,6 +47,14 @@
   @Cold
   Single<Project> loadProject(String projectId);
 
+
+  /**
+   * Loads the project terms from the remote data store. The return Single fails
+   * with if the terms are not found, or if the remote data store is not available.
+   */
+  @Cold
+  Single<Terms> loadTerms();
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/persistence/remote/RemoteDataStore.java:56]: terms->TermsOfService

File: gnd/src/main/java/com/google/android/gnd/persistence/remote/firestore/FirestoreDataStore.java [hunk 6]
```diff
@@ -65,6 +66,14 @@
         .subscribeOn(schedulers.io());
   }
 
+  @Cold
+  @Override
+  public  Single<Terms> loadTerms() {
+    return db.terms().getTerm().get()
+        .switchIfEmpty(Single.error(() -> new NotFoundException("No Terms of Service Founds")))
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/persistence/remote/firestore/FirestoreDataStore.java:73]: Consider returning a `Maybe` here and returning an error instead, since terms not being found is an expected/valid state in many cases.

File: gnd/src/main/java/com/google/android/gnd/persistence/remote/firestore/schema/TermsDocumentReference.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+public class TermsDocumentReference extends FluentDocumentReference {
+
+  TermsDocumentReference(DocumentReference ref) {
+    super(ref);
+  }
+
+  public TermsDocumentReference terms() {
+    return new TermsDocumentReference(reference());
+  }
+
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/persistence/remote/firestore/schema/TermsDocumentReference.java:34]: Nit: Extra LF

File: gnd/src/main/java/com/google/android/gnd/repository/TermsRepository.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+      RemoteDataStore remoteDataStore) {
+    this.remoteDataStore = remoteDataStore;
+
+  }
+
+  @Cold
+  public Flowable<Loadable<Terms>> getProjectTerms() {
+    return loadTermsFromRemote()
+        .doOnSubscribe(__ -> Timber.d("Loading terms from remote"))
+        .doOnError(err -> Timber.e(err, "Failed to load terms from remote"))
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/repository/TermsRepository.java:53]: "Not found" technically isn't an error, consider using `Timber.d` or removing.

File: gnd/src/main/java/com/google/android/gnd/ui/terms/TermsOfServiceFragment.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+      case LOADED:
+        binding.termsLoadingProgressBar.setVisibility(View.GONE);
+        binding.termsText.setVisibility(View.VISIBLE);
+        viewModel.setTermsTextView(projectTerms.value().get().getTerms());
+        break;
+      case NOT_FOUND:
+      case ERROR:
+        binding.termsLoadingProgressBar.setVisibility(View.GONE);
+        binding.termsText.setVisibility(View.VISIBLE);
+        viewModel.setTermsTextView("Some issue");
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/ui/terms/TermsOfServiceFragment.java:69]: Please use string resources, and set a more descriptive message. (e.g. "An error occurred while loading Terms of Service")

**Review by gino-m — CHANGES_REQUESTED** [review]
A few more suggestions. PTAL!

**Comment by gino-m** [comment]
@atalekar This PR is large so I didn't do a thorough review.. Could you also PTAL for better coverage?

**Comment by gino-m** [comment]
@HemanParbhakar Thanks for merging. Unit tests are still failing. You can run in Android Studio manually or by using `gradle compileStagingAndroidTestJavaWithJavac`.

File: gnd/src/main/java/com/google/android/gnd/model/Terms.java (hunk 2 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/model/Terms.java:21]: udpated

Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/model/Terms.java:21]: updated

File: gnd/src/main/java/com/google/android/gnd/model/Terms.java (hunk 3 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/model/Terms.java:23]: renamed to TermsOfService

File: gnd/src/main/java/com/google/android/gnd/persistence/local/LocalValueStore.java (hunk 4 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/persistence/local/LocalValueStore.java:81]: changed

File: gnd/src/main/java/com/google/android/gnd/persistence/remote/RemoteDataStore.java (hunk 5 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/persistence/remote/RemoteDataStore.java:56]: renamed

File: gnd/src/main/java/com/google/android/gnd/persistence/remote/firestore/schema/TermsDocumentReference.java (hunk 7 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/persistence/remote/firestore/schema/TermsDocumentReference.java:34]: removed

File: gnd/src/main/java/com/google/android/gnd/repository/TermsRepository.java (hunk 8 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/repository/TermsRepository.java:53]: replaced with Timber.d

File: gnd/src/main/java/com/google/android/gnd/ui/terms/TermsOfServiceFragment.java (hunk 9 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/ui/terms/TermsOfServiceFragment.java]: replaced with string resource

File: gnd/src/main/java/com/google/android/gnd/MainViewModel.java [hunk 10]
```diff
@@ -53,24 +58,34 @@
...
       AuthenticationManager authenticationManager,
       EphemeralPopups popups,
-      Schedulers schedulers) {
+      Schedulers schedulers,
+      LocalValueStore localValueStore) {
     this.projectRepository = projectRepository;
     this.featureRepository = featureRepository;
     this.navigator = navigator;
     this.popups = popups;
+    this.localValueStore = localValueStore;
```
[the remark this answers is not in the document]
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/MainViewModel.java:87]: now LocalValueStore is accessed using TermsOfServiceRepository.

File: gnd/src/main/java/com/google/android/gnd/MainViewModel.java [hunk 11]
```diff
@@ -129,6 +147,10 @@ private void onSignInStateChange(SignInState signInState) {
     }
   }
 
+  private void getProjectTerms(Loadable<Terms> projectTerms) {
```
[the remark this answers is not in the document]
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/MainViewModel.java]: this method update the state and which is later checked. As we discussed if there is not TermsOfService in remote db then redirect directly to the Main Page.

File: gnd/src/main/java/com/google/android/gnd/persistence/remote/firestore/FirestoreDataStore.java (hunk 6 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/persistence/remote/firestore/FirestoreDataStore.java:73]: NotFound Exception is later used to take the decision whether to redirect directly to the Main Page or not.

**Comment by HemanParbhakar** [comment]
@gino-m @atalekar this PR can be ignored latest PR is https://github.com/google/ground-android/pull/845

File: gnd/src/main/java/com/google/android/gnd/MainViewModel.java [hunk 12]
```diff
@@ -151,7 +173,16 @@ void onSignedOut() {
 
   private void onSignedIn() {
     hideProgressDialog();
-    navigator.navigate(HomeScreenFragmentDirections.showHomeScreen());
+    if (termsState.getValue() == LoadState.NOT_FOUND) {
+      localValueStore.setTermsAccepted(true);
+      navigator.navigate(HomeScreenFragmentDirections.showHomeScreen());
+    } else {
+      if (signInProgressDialogVisibility.getValue() == null) {
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/MainViewModel.java:180]: @HemanParbhakar Why is this necessary? Are you using the visibility of the sign in dialog as a proxy for something else?

**Comment by gino-m** [comment]
Closing in favor of updated PR #845.
